protocol/src/protocol.callHierarchy.proposed.md

```diff
@@ -0,0 +1,142 @@
+             */
+            dynamicRegistration?: boolean;
+        };
+}
+```
+
+_Server Capabilities_:
+
+```ts
+CallHierarchyServerCapabilities {
+    /**
+     * The server provides Call Hierarchy support.
+     */
+    callHierarchyProvider?: boolean | (TextDocumentRegistrationOptions & StaticRegistrationOptions);
+}
+```
+
+##### Call Hierarchy Request
+
+_Request_:
+
+The `textDocument/callHierarchy` request is sent from the client to the server to request the call hierarchy for a symbol at the given text document position.
+
+Returns a call hierarchy item for the requested call direction.
+
+* method: ‘textDocument/callHierarchy'
+* params: `CallHierarchyParams` defined as follows:
+
+```ts
+export interface CallHierarchyParams extends TextDocumentPositionParams {
+    resolve?: number;
+    direction?: 'incoming' | 'outgoing';
+}
+```
+
+_Response_:
+
+The server will send a `CallHierarchyItem` object containing the information about the targeted symbol. The item will be undefined, if no such symbol is found.
+
+The item is _unresolved_ if the lists of callers and callees are undefined. Unresolved items can be resolved via `callHierarchy/resolve` requests.
+
+The resolved item includes callers or callees again of type `CallHierarchyItem`. The caller/callee object provide the actual location of the call (`CallHierarchyItem.callLocation`).
+
+* result: `CallHierarchyItem` defined as follows:
+
+```ts
+export interface CallHierarchyItem {
+    /**
+     * The name of the symbol targeted by the call hierarchy request.
+     */
+    name: string;
+    /**
+     * More detail for this symbol, e.g the signature of a function.
+     */
+    detail?: string;
+    /**
+     * The kind of this symbol.
+     */
+    kind: SymbolKind;
+    /**
+     * URI of the document containing the symbol.
+     */
+    uri: string;
+    /**
+     * The range enclosing this symbol not including leading/trailing whitespace but everything else
+     * like comments. This information is typically used to determine if the the clients cursor is
+     * inside the symbol to reveal in the symbol in the UI.
+     */
+    range: Range;
+    /**
+     * The range that should be selected and revealed when this symbol is being picked, e.g the name of a function.
+     * Must be contained by the the `range`.
+     */
+    selectionRange: Range;
+
+    /**
+     * The actual location of the call.
+     *
+     * Must be defined in resolved callers/callees.
+     */
+    callLocation?: Location;
+
+    /**
+     * List of incoming calls.
+     *
+     * *Note*: The items is _unresolved_ if `callers` and `callees` is undefined.
+     */
+    callers?: CallHierarchyItem[];
+    /**
+     * List of outgoing calls.
+     *
+     * *Note*: The items is _unresolved_ if `callers` and `callees` is undefined.
+     */
+	callees?: CallHierarchyItem[];
+	/**
+     * Optional data to identify an item in a resolve request.
+     */
+    data?: any;
```

Comment:
Maybe, we could add a `recursive?: boolean` property. `true` if the method/function represents a recursion call. Thoughts?

Reply:
A `deprecated?: boolean` could be useful too.

Reply:
on `recursive?: boolean`: 
> `true` if the method/function represents a recursion call

I recall discussing that while prototyping the call hierarchy feature in Theia, and we came to the conclusion to leave the detection of cycles to the clients. 

Detecting cycles (of random length) might be quite hard when implementing the lazy-loading strategy?